### PR TITLE
Multi Agent Training

### PR DESCRIPTION
**Pull request by rudrasohan**
What?
- Added Support for multi-agent training following the [BAIR](https://bair.berkeley.edu/blog/2018/12/12/rllib/) guidelines.

- Supports training with rllib.

Testing:
- run `python train_rllib_multi_agent.py` 

NOTE:
- Although the training is running, the process is very CPU intensive.
- Due to this, it crashes sometimes in my local machine.
- @Santara could you check the code once on your machine.

### CODE REVIEW

File: MADRaS/envs/gym_madras_v2.py [hunk 1]
```diff
@@ -383,11 +384,17 @@ def step(self, action):
...
             proc.join()
+        
+        done_check = False
+
         for agent in self.agents:
             next_obs[agent] = return_dict[agent][0]
             reward[agent] = return_dict[agent][1]
             done[agent] = return_dict[agent][2]
+            if (done[agent] == True):
+                done_check = True
```
Comment (Santara) [inline MADRaS/envs/gym_madras_v2.py:401]: Maybe add a comment that this only supports the case in which all the agents are forced to terminate their episodes when one of them hits `done`?

File: MADRaS/agents/rllib/train_rllib_multi_agent.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    ppo_trainer = PPOTrainer(
+        env="Madras-v0",
+        config={
+            "multiagent": {
+                "policies": policies,
+                "policy_mapping_fn": policy_mapping_fn,
+                #"policies_to_train": ["ppo_policy_1","ppo_policy_2"],
+            },
+            # disable filters, otherwise we would need to synchronize those
+            # as well to the DQN agent
```
Comment (Santara) [inline MADRaS/agents/rllib/train_rllib_multi_agent.py]: remove unnecessary comments

File: MADRaS/agents/rllib/train_rllib_multi_agent.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    # info:
+    #   policy_reward_mean:
+    #     dqn_policy: X
+    #     ppo_policy: Y
+    for i in range(args.num_iters):
+        print("== Iteration", i, "==")
+
+        # improve the DQN policy
+        #print("-- DQN --")
+        #print(pretty_print(dqn_trainer.train()))
```
Comment (Santara) [inline MADRaS/agents/rllib/train_rllib_multi_agent.py]: remove comments

File: MADRaS/agents/rllib/train_rllib_multi_agent.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    # disable DQN exploration when used by the PPO trainer
+    # ppo_trainer.workers.foreach_worker(
+    #     lambda ev: ev.for_policy(
+    #         lambda pi: pi.set_epsilon(0.0), policy_id="dqn_policy"))
+
+    # You should see both the printed X and Y approach 200 as this trains:
+    # info:
+    #   policy_reward_mean:
+    #     dqn_policy: X
+    #     ppo_policy: Y
```
Comment (Santara) [inline MADRaS/agents/rllib/train_rllib_multi_agent.py]: remove unnecessary comments

File: MADRaS/agents/rllib/train_rllib_multi_agent.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        #print("-- DQN --")
+        #print(pretty_print(dqn_trainer.train()))
+
+        # improve the PPO policy
+        print("-- PPO --")
+        print(pretty_print(ppo_trainer.train()))
+
+        # swap weights to synchronize
+        #dqn_trainer.set_weights(ppo_trainer.get_weights(["ppo_policy"]))
+        #ppo_trainer.set_weights(dqn_trainer.get_weights(["dqn_policy"]))
```
Comment (Santara) [inline MADRaS/agents/rllib/train_rllib_multi_agent.py]: add a flag as an option to do this?

File: MADRaS/agents/rllib/train_rllib_multi_agent.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    obs_space_1 = env.agents["MadrasAgent_0"].observation_space
+    act_space_1 = env.agents["MadrasAgent_0"].action_space
+    obs_space_2 = env.agents["MadrasAgent_1"].observation_space
+    act_space_2 = env.agents["MadrasAgent_1"].action_space
+    # You can also have multiple policies per trainer, but here we just
+    # show one each for PPO and DQN.
+    policies = {
+        "ppo_policy_1": (PPOTFPolicy, obs_space_1, act_space_1, {}),
+        "ppo_policy_2": (PPOTFPolicy, obs_space_2, act_space_2, {}),
+    }
```
Comment (Santara) [inline MADRaS/agents/rllib/train_rllib_multi_agent.py]: Remove hardcoding by using lists:
```
obs_spaces, action_spaces = [], []
for agent in env.agents:
  obs_spaces.append(agent.observation_space)
  action_spaces.append(agent.action_space)

policies = {"ppo_policy_{}".format(i) : (PPOTFPolicy, obs_spaces[i], action_spaces[i]) for i in range(len(env.agents.keys()))}
```

File: MADRaS/agents/rllib/train_rllib_multi_agent.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+    def policy_mapping_fn(agent_id):
+        if agent_id == "MadrasAgent_0":
+            return "ppo_policy_1"
+        else:
+            return "ppo_policy_2"
```
Comment (Santara) [inline MADRaS/agents/rllib/train_rllib_multi_agent.py]: Remove hardcoding like the comment above

File: MADRaS/envs/gym_madras_v2.py [hunk 8]
```diff
@@ -256,7 +257,7 @@ def increment_step(self):
             self.step_num += 1
 
 
-class MadrasEnv(gym.Env):
+class MadrasEnv(gym.Env, MultiAgentEnv):
```
Comment (Santara) [inline MADRaS/envs/gym_madras_v2.py]: This makes MADRaS inherently dependent on RLLib. Do we want that? Such a tight integration with RLLib may result in incompatibility with other RL Frameworks like RLPyT. I was thinking of a derived class of MadrasEnv that adds `MultiAgentEnv` as a parent.

File: setup.py [hunk 9]
```diff
@@ -8,4 +8,5 @@
                         'pyyaml',
                         'pTable',
                         'matplotlib',
-                        'mpi4py'])
+                        'mpi4py',
+                        'ray[rllib]'])
```
Comment (Santara) [inline setup.py]: Is it a good idea to make MADRaS dependent on RLLib? I suggest we remove this dependency.

**Review by Santara — COMMENTED** [review]
Thanks for writing wrapping MADRaS around the `MultiAgentEnv` of RLLib. I have a few questions/comments - please have a look. :)
Here is a summary:
0. Everything runs! :+1: both with visualization on and off.
1. I suggest we make a derived class of `MadrasEnv` called `MadrasEnvRllib` that inherits `MultiAgentEnv`. Making the main environment dependent on RLLib may break compatibilty with other RL libraries.
2. Rollouts are extremely slow and the learning agents face timeout socket errors very frequently which i think slows down rollouts.

File: setup.py (hunk 9 above)
Comment (rudrasohan) [reply setup.py]: I too think that dependence on RLLib will cause troubles. I'll try and find a workaround.

File: MADRaS/agents/rllib/train_rllib_multi_agent.py (hunk 6 above)
Comment (rudrasohan) [reply MADRaS/agents/rllib/train_rllib_multi_agent.py]: Awesome

File: MADRaS/agents/rllib/train_rllib_multi_agent.py (hunk 5 above)
Comment (rudrasohan) [reply MADRaS/agents/rllib/train_rllib_multi_agent.py]: This is not required as of now as we are only using ppo.

File: MADRaS/agents/rllib/train_rllib_multi_agent.py [hunk 10]
```diff
@@ -5,6 +5,23 @@
 from ray.rllib.agents.ppo.ppo_policy import PPOTFPolicy
 from ray.tune.logger import pretty_print
 from ray.rllib.env.multi_agent_env import MultiAgentEnv
+from MADRaS.envs.gym_madras_v2 import MadrasEnv
+
+
+class MADRaSRLLib(MultiAgentEnv):
```
Comment (Santara) [inline MADRaS/agents/rllib/train_rllib_multi_agent.py]: Lets stick to our convention of Camel Case for class names and make `MADRaSRLLib` --> `MadrasRllib`

Comment (Santara) [inline MADRaS/agents/rllib/train_rllib_multi_agent.py]: Lets stick to our convention of Camel Case for class names and make `MADRaSRLLib` --> `MadrasRllib`

Comment (Santara) [inline MADRaS/agents/rllib/train_rllib_multi_agent.py]: Add `MadrasEnv` as a parent class to inherit `reset()` and `step()`

File: MADRaS/agents/rllib/train_rllib_multi_agent.py [hunk 11]
```diff
@@ -13,61 +30,36 @@
...
-    act_space_2 = env.agents["MadrasAgent_1"].action_space
+    env = MADRaSRLLib()
+
+    obs_spaces, action_spaces = [], []
+    for agent in env.agents:
+        obs_spaces.append(env.agents[agent].observation_space)
+        action_spaces.append(env.agents[agent].action_space)
+
+    policies = {"ppo_policy_{}".format(i) : (PPOTFPolicy, obs_spaces[i], action_spaces[i], {}) for i in range(env.num_agents)}
     # You can also have multiple policies per trainer, but here we just
```
Comment (Santara) [inline MADRaS/agents/rllib/train_rllib_multi_agent.py]: we dont need this comment too

**Review by Santara — CHANGES_REQUESTED** [review]
A few more comments... :)

**Comment by Santara** [comment]
Thank you for responding to my comments. Training also happens smoothly now. Could you add the name of the agent to the logging statements like this: https://github.com/Santara/MADRaS/blob/9e28bf1099d9516b4c6674521373a59ac54e8426/MADRaS/utils/done_manager_v2.py#L91 
For multi-agent training, it might be helpful to identify which of the agents face what kind of done condition more often.

**Comment by rudrasohan** [comment]
Done Feature
image

The agent name followed by the reasons for done. All the reasons will be presented in the form of a list.

**Comment by Santara** [comment]
Thanks for implementing the feature, @rudrasohan! Its neat but I was thinking about making `name` an attribute of the `DoneManager` and `MadrasDone` classes. This can then be added to the original done statements to make them look like:
```
[MadrasAgent-0]: Done : Episode terminated because agent collided.
[MadrasAgent-1]: Done : Episode terminated because agent collided.
```
This will print one logging message per agent for Done instead of two in your implementation.

**Comment by rudrasohan** [comment]
Screenshot from 2019-12-21 23-57-37
Screenshot from 2019-12-21 23-43-35

Reformatted. Currently, each `DoneManager` has and `agent_id` attr and each `MadrasDone` has a `reason` attr.

**Comment by Santara** [comment]
LGTM, Approval.

## Known issue:
A few times, agents fail to connect to the TORCS server ever and get stuck in an infinite loop of timeouts. The last time this happened, I had opened Firefox the very moment an episode had ended and TORCS had to be relaunched. The port that TORCS server was running on might have been occupied by Firefox and the server never came up. All the agents kept looping in timeout errors.

## Probable fix:
`MadrasEnv.reset()` launches TORCS on the same port all the time. It should first test if the port is available before trying to launch TORCS there. This exercise should be repeated every time TORCS is relaunched. This also means that we should revise the port information that we send to the `MadrasAgent`s and the `traffic_manager`. We can do this by re-instantiating the traffic manager and MadrasAgents at every reset.

**Comment by rudrasohan** [comment]
I shall incorporate this in this in a separate PR.

**Pull request merged**